api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -314,13 +314,21 @@ message SubjectAltNameMatcher {
     DNS = 2;
     URI = 3;
     IP_ADDRESS = 4;
+    OTHER_NAME = 5;
   }
 
   // Specification of type of SAN. Note that the default enum value is an invalid choice.
   SanType san_type = 1 [(validate.rules).enum = {defined_only: true not_in: 0}];
 
   // Matcher for SAN value.
   type.matcher.v3.StringMatcher matcher = 2 [(validate.rules).message = {required: true}];
+
+  // OID Value which is required if OTHER_NAME SAN type is used.
+  // For example, UPN OID is 1.3.6.1.4.1.311.20.2.3
+  // (Reference: http://oid-info.com/get/1.3.6.1.4.1.311.20.2.3).
+  //
+  // If set for SAN types other than OTHER_NAME, it will be ignored.
+  string oid = 3;
```

Comment:
Shouldn't we just use the matcher field above like the rest do and if configured for OTHER_NAME use that for matching?

Reply:
I could be misunderstanding your question, but it seems to me like you are asking why are we using a separate field OID for the OtherName type to match. OtherName requires two values according to RFC 5280 - an object identifier(OID) to confine it to a particular type(eg: UPN, Kerberos Principal Name, etc.) and the actual value. 

Validating OID is required to make sure our expectation of OtherName(UPN) with value X doesn't match against OtherName(Kerberos Principal Name) with value X, hence the need for the separate field and the matching logic for it. The matcher field above will work for the actual OtherName value as it does with other SAN types.

Reply:
I also had some initial discussions around this with @ggreenway here - #34358

Reply:
> According to RFC 5280, OtherName requires two values: an object identifier (OID) to specify the type (e.g., UPN, Kerberos Principal Name, etc.) and the actual value.
> 

Checking the code, it seems like the `value` is ignored. And the value can be binary data. Should we have a field for the value and it can be `google.protobuf.BytesValue` type?

Reply:
Value is not ignored, but rather being currently matched using the matcher field, which confines it to only string data. One of the test cases is a cert containing the SAN OtherName(UPN): server1.example.com  (https://github.com/envoyproxy/envoy/pull/34471/files#diff-db3a1575b8e3dcb50ed0389a327546c565e69527c390b231ecb0e1a94558c3a6). Here we can do a string regex match against server1.example.com. 

For binary values, as you suggested, we can add an additional field and let the user define either the matcher field or the binary_value field (binary value will be a strict comparison)

Reply:
I think the existing value matcher is good enough; for non-string data, you can use the `custom` type from https://www.envoyproxy.io/docs/envoy/latest/xds/type/matcher/v3/string.proto.html.

Reply:
Can we update the API then we can continue the review? Thanks.

/wait

Reply:
My understanding was that the API changes was going to remain as is, apart from the doc changes. Am I missing something here?

Reply:
OK I misunderstood, sorry. OK this looks fine to me, thanks.